autopush_rs/src/client.rs

```diff
@@ -112,747 +211,837 @@ impl ClientFlags {
-                        self.data.webpush.as_mut().unwrap().stats.nacks += 1;
-                        ClientState::Await
-                    }
-                    Either::A(ClientMessage::Ack { updates }) => self.data.process_acks(updates),
-                    Either::B(ServerNotification::Notification(notif)) => {
-                        let webpush = self.data.webpush.as_mut().unwrap();
-                        if notif.ttl != 0 {
-                            webpush.unacked_direct_notifs.push(notif.clone());
-                        }
-                        debug!("Got a notification to send, sending!");
-                        ClientState::FinishSend(
-                            Some(ServerMessage::Notification(notif)),
-                            Some(Box::new(ClientState::Await)),
-                        )
-                    }
-                    Either::B(ServerNotification::CheckStorage) => {
-                        let webpush = self.data.webpush.as_mut().unwrap();
-                        webpush.flags.include_topic = true;
-                        webpush.flags.check = true;
-                        ClientState::Await
-                    }
-                    Either::B(ServerNotification::Disconnect) => {
-                        debug!("Got told to disconnect, connecting client has our uaid");
-                        ClientState::ShutdownCleanup(Some("Repeat UAID disconnect".into()))
-                    }
-                    _ => return Err("Invalid message".into()),
-                }
-            }
-            ClientState::ShutdownCleanup(ref mut err) => {
-                debug!("State: ShutdownCleanup");
-                if let Some(err_obj) = err.take() {
-                    let mut error = err_obj.to_string();
-                    for err in err_obj.iter().skip(1) {
-                        error.push_str("\n");
-                        error.push_str(&err.to_string());
-                    }
-                    debug!("Error for shutdown of {}: {}", host, error);
-                };
-                self.data.shutdown();
-                ClientState::Done
-            }
-            ClientState::Done => {
-                // We don't expect this to actually run, as this state will exit
-                // the transition. Included for exhaustive matching.
-                debug!("State: Done");
-                ClientState::Done
-            }
-        };
-        Ok(next_state.into())
-    }
+    #[state_machine_future(transitions(UnAuthDone))]
+    AwaitSessionComplete {
+        auth_state_machine: AuthClientStateFuture<T>,
+        srv: Rc<Server>,
+        user_agent: String,
+        webpush: Rc<RefCell<WebPushClient>>,
+    },
+
+    #[state_machine_future(ready)]
+    UnAuthDone(()),
+
+    #[state_machine_future(error)]
+    UnAuthClientStateError(Error),
 }
 
-impl<T> ClientData<T>
+impl<T> PollUnAuthClientState<T> for UnAuthClientState<T>
```

Comment:
Add a comment noting that Impl and method names have a required formula? This might be easy to screw up in the future?